### PR TITLE
Added def type and other improvements

### PR DESCRIPTION
**Pull request by jdconrad**
* Significant improvements to the definition including lower bounding of type pairs.
* Significant improvments to type promotion including separate promotion for conditionals leading to less coupling between nodes in the tree.
* Added a dynamic (def) type that can be any type and will do runtime checking for class types.
* Overloaded operators work with def types.
* General clean up of writer.
* Removed cat .. operator in favor of overloaded + operator.

### CODE REVIEW

File: src/test/resources/rest-api-spec/test/plan_a/20_scriptfield.yaml [hunk 1]
```diff
@@ -19,7 +19,7 @@ setup:
                 script_fields:
                     bar:
                         script: 
-                            inline: "return ((List)((Map<String,Object>)input.get(\"doc\")).get(\"foo\")).get(0) .. input.get(\"x\");"
+                            inline: "return input.get(\"doc\").get(\"foo\").get(0) + input.get(\"x\");"
                             lang: plan-a
```
Comment (rmuir) [inline src/test/resources/rest-api-spec/test/plan_a/20_scriptfield.yaml]: huge improvement. Do you mind improving 30_search.yaml too?

**Comment by rmuir** [comment]
looks good, awesome progress

File: src/test/resources/rest-api-spec/test/plan_a/20_scriptfield.yaml (hunk 1 above)
Comment (jdconrad) [inline src/test/resources/rest-api-spec/test/plan_a/20_scriptfield.yaml]: Done.

**Comment by jdconrad** [comment]
Thanks for the review!

**Pull request merged**